### PR TITLE
Fix issue when there are multiple candidates for edit_distance_with_substrings

### LINKED ISSUE #109291: Poor suggestion of alternative function when a name is incorrect
### Code

```Rust
println!("Custom backtrace: {}", std::backtrace::Backtrace::forced_capture());
```

### Current output

```Shell
error[E0599]: no function or associated item named `forced_capture` found for struct `Backtrace` in the current scope
  --> src/line_parser.rs:19:65
   |
19 |     println!("Custom backtrace: {}", std::backtrace::Backtrace::forced_capture());
   |                                                                 ^^^^^^^^^^^^^^
   |                                                                 |
   |                                                                 function or associated item not found in `Backtrace`
   |                                                                 help: there is an associated function with a similar name: `capture`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `difftastic` due to previous error
```

### Desired output

It would be better to suggest `force_capture` rather than `capture`, as that was the function I was looking for.

### Rationale and extra context

_No response_

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #109291

### CODE REVIEW

**Comment by OliverMD** [comment]
@chenyukang Nice! fwiw this seems like a good idea to me

File: compiler/rustc_span/src/edit_distance.rs [hunk 1]
```diff
@@ -198,12 +199,27 @@ fn find_best_match_for_name_impl(
...
-                dist = d - 1;
-                best = Some(*c);
+                if use_substring_score {
+                    dist = d;
+                    next_candidates.push(*c);
+                    best = Some(*c);
+                } else {
+                    dist = d - 1;
+                    best = Some(*c);
+                }
```
Comment (cjgillot) [inline compiler/rustc_span/src/edit_distance.rs:209]: Could you explain why we introduce a different behaviour here?
In particular in the assignment to `dist`?

File: compiler/rustc_span/src/edit_distance.rs [hunk 2]
```diff
@@ -190,6 +190,7 @@ fn find_best_match_for_name_impl(
 
     let mut dist = dist.unwrap_or_else(|| cmp::max(lookup.len(), 3) / 3);
     let mut best = None;
+    let mut next_candidates = vec![];
```
Comment (cjgillot) [inline compiler/rustc_span/src/edit_distance.rs:193]: Could you add a comment stating what you put in this vector?

File: compiler/rustc_span/src/edit_distance.rs [hunk 3]
```diff
@@ -198,12 +199,27 @@ fn find_best_match_for_name_impl(
...
+                } else {
+                    dist = d - 1;
+                    best = Some(*c);
+                }
             }
             None => {}
         }
     }
+
+    if next_candidates.len() > 1 {
```
Comment (cjgillot) [inline compiler/rustc_span/src/edit_distance.rs:215]: Could you add a comment stating what happens here?
Does this need an `debug_assert!(use_substring_score)` to emphasize that only happens in that case?

Comment (chenyukang) [reply compiler/rustc_span/src/edit_distance.rs:221]: yes, added the assertion in code.

File: compiler/rustc_span/src/edit_distance.rs (hunk 1 above)
Comment (chenyukang) [reply compiler/rustc_span/src/edit_distance.rs]: updated it!

File: compiler/rustc_span/src/edit_distance.rs (hunk 3 above)
Comment (cjgillot) [reply compiler/rustc_span/src/edit_distance.rs:221]: r=me with a comment explaining this branch. Example (to be adapted)
```suggestion
    //  We have a tie among several candidates, try to select the best among them ignoring substrings.
    if next_candidates.len() > 1 {
```

**Pull request merged**